[ML] Text/Log categorization multi-bucket aggregation (#71752)
elastic/elasticsearch: docs/reference/aggregations/bucket/categorize-text-aggregation.asciidoc

[[search-aggregations-bucket-categorize-text-aggregation]]
=== Categorize text aggregation
++++
<titleabbrev>Categorize text</titleabbrev>
++++

experimental::[]

A multi-bucket aggregation that groups semi-structured text into buckets. Each `text` field is re-analyzed
using a custom analyzer. The resulting tokens are then categorized creating buckets of similarly formatted
text values. This aggregation works best with machine generated text like system logs.

WARNING: Re-analyzing _large_ result sets will require a lot of time and memory. This aggregation should be
         used in conjunction with <<async-search, Async search>>.

[[bucket-categorize-text-agg-syntax]]
==== Parameters

`field`::
(Required, string)
The semi-structured text field to categorize.

`max_children`::
(Optional, integer, default: `100`)
The maximum number of unique tokens at any given layer of the tokenization tree.

> Must be larger than 1. The smaller the value, the more broad the text categories.

Reviewer: What does "broad" mean in this sentence?

Author: @droberts195 good point. I am unsure how to communicate more exact vs more widely encompassing categories. 

Fewer children means at each layer of the tree we will start using `*` more quickly. So, we may merge categories more quickly and will definitely use less memory. 

I will think on how to better to explain this.

> Larger values may cause the aggregation to more memory and run more slowly

Reviewer: ```suggestion
Larger values may cause the aggregation to use more memory and run more slowly.
```

(The word "use" is also missing from this sentence in the description immediately underneath.)

`max_depth`::
(Optional, integer, default: `5`)
The maximum number of tokens matched on before attempting to merge categories.

> Larger values may cause the aggregation to more memory and run more slowly.

Reviewer: I wonder if this argument should be renamed.  Reading the description it's not really clear what the effect of making the number bigger or smaller will be on the categories.  And also the "attempting to" part suggests that setting it to, say, 7, won't guarantee you never get two categories with more than 7 matching tokens.

Maybe something along the lines of `merge_token_count`, `merge_strength`, `merge_propensity`. (Although none of these are great they might inspire someone to think of a really good name.)

And add a sentence to the description like, "In general, the lower this number the fewer categories you will get."

`similarity_threshold`::
(Optional, double, default: `0.5`)

> The minimum percentage of tokens that must match for text to be added to the

Reviewer: Currently it's a proportion, not a percentage.  Maybe it _should_ be a percentage.  So either change the docs or actually make it a percentage and divide the setting value by 100 internally before use.  Maybe there is prior art in other aggregations' settings to suggest which way to go to be most consistent.

Author: A percentage has the added benefit of not allowing values like `0.15789291058910` which don't add much...

A percentage would also match up with how we handle things in data frame analytics. Though I don't know about prior art in the aggs.

category bucket.
Must be between 0.1 and 1.0. The larger the value the more restrictive the log categories.
Larger values may increase memory usage.

> `categorization_filters`::

Reviewer: I think it would be good to let people customize the whole analyzer if they want to.  When configuring a categorization job users can choose:

1. Entirely default tokenization and filtering
2. Largely default tokenization and filtering, but with the addition of categorization filters (which are basically pattern char filters that get injected into the default analyzer)
3. Totally custom analyzer

Since you are reusing the `CategorizationAnalyzerConfig` class most of the complexity in having these 3 options will be in code that can be reused.

The parsing bit you'll have to duplicate is in https://github.com/elastic/elasticsearch/blob/15933c0b8ea199de5de3faa8044e8d322138094e/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/AnalysisConfig.java#L93-L95

And then you need a check that `categorization_filters` and `categorization_analyzer` are not _both_ specified together.

But it doesn't seem _that_ hard to add now.

I was going to say it could be added as a followup but I am wondering if the effort of getting master and 7.x to talk to each other with master having the more complex options and 7.x not will actually be more work than just adding to this PR.  See what you think - if you'd rather add it in a followup PR and you don't think the (temporary) BWC logic will be a nightmare then fair enough.

Author: > I think it would be good to let people customize the whole analyzer if they want to. 

Agreed, I can add that.

[[search-aggregations-bucket-categorize-text-aggregation]]
=== Categorize text aggregation
++++
<titleabbrev>Categorize text</titleabbrev>
++++

experimental::[]

A multi-bucket aggregation that groups semi-structured text into buckets. Each `text` field is re-analyzed
using a custom analyzer. The resulting tokens are then categorized creating buckets of similarly formatted
text values. This aggregation works best with machine generated text like system logs.

WARNING: Re-analyzing _large_ result sets will require a lot of time and memory. This aggregation should be
         used in conjunction with <<async-search, Async search>>.

[[bucket-categorize-text-agg-syntax]]
==== Parameters

`field`::
(Required, string)
The semi-structured text field to categorize.

`max_children`::
(Optional, integer, default: `50`)
The maximum number of unique tokens at any given layer of the tokenization tree.

> Must be larger than 1. Smaller values use less memory and create wider text categories.
> Larger values will use more memory and create narrower categories.

Reviewer: ```suggestion
Must be larger than 1. Smaller values use less memory and create fewer categories.
Larger values will use more memory and create more categories.
```

`max_depth`::
(Optional, integer, default: `5`)
The maximum number of tokens matched on before attempting to merge categories.

> Larger values will use more memory and create narrower categories.

Author: `max_tokens` would be tricky as well as it may assume that "this is the max size log we match", which isn't true. Its just the number of token vectors.

I am still thinking on a better name

`similarity_threshold`::
(Optional, integer, default: `50`)
The minimum percentage of tokens that must match for text to be added to the
category bucket.
Must be between 1 and 100. The larger the value the narrower the categories.
Larger values will increase memory usage and create narrower categories.

`categorization_filters`::
(Optional, array of strings)
This property expects an array of regular expressions. The expressions
are used to filter out matching sequences from the categorization field values.
You can use this functionality to fine tune the categorization by excluding
sequences from consideration when categories are defined. For example, you can
exclude SQL statements that appear in your log files. This
property cannot be used at the same time as `categorization_analyzer`. If you
only want to define simple regular expression filters that are applied prior to
tokenization, setting this property is the easiest method. If you also want to
customize the tokenizer or post-tokenization filtering, use the
`categorization_analyzer` property instead and include the filters as

> `pattern_replace` character filters.

Author: @nik9000 true, we do supply some examples in our ML documentation.

`categorization_analyzer`::

> (Optional, (object or string)

Reviewer: ```suggestion
(Optional, object or string)
```

++++

experimental::[]

A multi-bucket aggregation that groups semi-structured text into buckets. Each `text` field is re-analyzed
using a custom analyzer. The resulting tokens are then categorized creating buckets of similarly formatted
text values. This aggregation works best with machine generated text like system logs.

WARNING: Re-analyzing _large_ result sets will require a lot of time and memory. This aggregation should be
         used in conjunction with <<async-search, Async search>>. Additionally, you may consider
         using the aggregation as a child of either the <<search-aggregations-bucket-sampler-aggregation,sampler>> or
         <<search-aggregations-bucket-diversified-sampler-aggregation,diversified sampler>> aggregation.
         This will typically improve speed and memory use.

NOTE: If you have considerable memory allocated to your JVM but are receiving circuit breaker exceptions from this
      aggregation, you may be attempting to categorize text that is poorly formatted for categorization. Consider
      adding `categorization_filters` or running under <<search-aggregations-bucket-sampler-aggregation,sampler>> or

>       <<search-aggregations-bucket-diversified-sampler-aggregation,diversified sampler>> to explore the created categories.

Author: > It'd be nice if there was some signal about which tokens were consuming the most memory. That's a "for later" thing, I think.

This would be pretty difficult. That would require some sort of pattern recognition inside of a given token position. This would be neat, but non-trivial.

>       <<search-aggregations-bucket-diversified-sampler-aggregation,diversified sampler>> to explore the created categories.

Reviewer 2: Yeah. Would you be ok moving this lower into the page?

The number of buckets to return.

`min_doc_count`::
(Optional, integer)
The minimum number of documents for a bucket to be returned to the results.

`shard_min_doc_count`::
(Optional, integer)
The minimum number of documents for a bucket to be returned from the shard before
merging.

==== Basic use

Example:

[source,console]
--------------------------------------------------
POST log-messages/_search?filter_path=aggregations
{
  "aggs": {
    "categories": {
      "categorize_text": {
        "field": "message"
      }
    }
  }
}
--------------------------------------------------
// TEST[setup:categorize_text]

Response:

[source,console-result]
--------------------------------------------------
{
  "aggregations" : {
    "categories" : {
      "buckets" : [
        {
          "doc_count" : 3,
          "key" : "Node shutting down"
        },
        {
          "doc_count" : 1,
          "key" : "Node starting up"
        },
        {
          "doc_count" : 1,
          "key" : "User foo_325 logging on"
        },
        {
          "doc_count" : 1,
          "key" : "User foo_864 logged off"
        }
      ]
    }
  }
}
--------------------------------------------------

> Here is an example using `categorization_filters`

Reviewer 2: Should my example be made to work? Maybe by running the regexes before tokenization? It seems useful from where I sit. I just don't know for sure.

> Here is an example using `categorization_filters`

Author: > Should my example be made to work? Maybe by running the regexes before tokenization? It seems useful from where I sit. I just don't know for sure.

This makes it significantly different than the ML tokenization that has proven itself over the years. I would prefer not doing this out of the box. 

If a custom analyzer + tokenizer can support this, then the user can work around it by providing their own.
